### PR TITLE
feat: add firestore database

### PR DESCRIPTION
**Pull request by ViktorMank**

### CODE REVIEW

File: src/store/slices/initSlice/initSlice.ts [hunk 1]
```diff
@@ -1,5 +1,7 @@
 import { getAuth, onAuthStateChanged } from 'firebase/auth';
 
+import { getUserData } from '../../../integrations/firebase/usersDatabase';
```
Comment (thelastandrew) [inline src/store/slices/initSlice/initSlice.ts:3]: fix path 'integrations/firebase/usersDatabase'

File: src/store/slices/signInSlice/signInSlice.ts [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import { getUserData, setUserData } from '../../../integrations/firebase/usersDatabase';
```
Comment (thelastandrew) [inline src/store/slices/signInSlice/signInSlice.ts:1]: fix path 'integrations/firebase/usersDatabase'

File: src/store/slices/signInSlice/signInSlice.ts [hunk 3]
```diff
@@ -30,19 +32,24 @@ export const signInSlice: GenericStateCreator<BoundStore> = (set, get) => ({
...
+              setUserData({
+                id: user.uid,
+                phone: user.phoneNumber,
+                name: user.displayName,
+                image: user.photoURL,
+                role: 'user',
+              });
+            }
+          } catch (error) {
+            console.error(error);
```
Comment (thelastandrew) [inline src/store/slices/signInSlice/signInSlice.ts:48]: delete console.log, set error to the corresponding property of the slice (see line 10)

File: src/store/slices/initSlice/initSlice.ts [hunk 4]
```diff
@@ -8,6 +10,7 @@ export interface AppUser {
   phone: Nullable<string>;
   name: Nullable<string>;
   image: Nullable<string>;
+  role: string;
```
Comment (thelastandrew) [inline src/store/slices/initSlice/initSlice.ts:13]: role: 'user' | 'admin'

File: src/store/slices/signInSlice/signInSlice.ts [hunk 5]
```diff
@@ -30,19 +32,24 @@ export const signInSlice: GenericStateCreator<BoundStore> = (set, get) => ({
...
+          try {
+            const userData = await getUserData(user);
+
+            if (!userData) {
+              setUserData({
+                id: user.uid,
+                phone: user.phoneNumber,
+                name: user.displayName,
+                image: user.photoURL,
+                role: 'user',
```
Comment (thelastandrew) [inline src/store/slices/signInSlice/signInSlice.ts:44]: role: user.role | 'user'

File: src/store/slices/signInSlice/signInSlice.ts [hunk 6]
```diff
@@ -30,19 +32,24 @@ export const signInSlice: GenericStateCreator<BoundStore> = (set, get) => ({
...
+              });
+            }
+          } catch (error) {
+            console.error(error);
+          }
           set((state) => ({
             ...state,
-            user: {
```
Comment (thelastandrew) [inline src/store/slices/signInSlice/signInSlice.ts:35]: we still need to set current user to the store

File: src/store/slices/signInSlice/signInSlice.ts [hunk 7]
```diff
@@ -30,19 +32,24 @@ export const signInSlice: GenericStateCreator<BoundStore> = (set, get) => ({
...
-              id: user.uid,
-              phone: user.phoneNumber,
-              name: user.displayName,
-              image: user.photoURL,
-            },
             isAuth: true,
-            signIn: {
-              ...state.signIn,
-              isFetching: false,
```
Comment (thelastandrew) [inline src/store/slices/signInSlice/signInSlice.ts:44]: and to set fetching to false

File: src/store/slices/initSlice/initSlice.ts [hunk 8]
```diff
@@ -32,18 +35,27 @@ export const initSlice: GenericStateCreator<BoundStore> = (set, get) => ({
...
+                phone: userData.phone,
+                name: userData.name,
+                image: userData.image,
+                role: userData.role,
+              },
+              isAuth: true,
+            }));
+          }
+        } catch (error) {
+          console.error(error);
```
Comment (thelastandrew) [inline src/store/slices/initSlice/initSlice.ts:57]: remove console.log,
declare nullable property 'error',
set caught error to this property,
then we can handle this 'error' in React component

**Review by thelastandrew — CHANGES_REQUESTED** [review]
.

File: src/store/slices/initSlice/initSlice.ts (hunk 4 above)
Comment (remadyt) [reply src/store/slices/initSlice/initSlice.ts:13]: create enum for this

File: src/store/slices/initSlice/initSlice.ts (hunk 8 above)
Comment (remadyt) [reply src/store/slices/initSlice/initSlice.ts:57]: maybe use notification here? and all such places

File: src/store/slices/signInSlice/signInSlice.ts (hunk 5 above)
Comment (remadyt) [reply src/store/slices/signInSlice/signInSlice.ts:44]: dont handle role from frontend. need set roles only in db

and why u desctruring this object? setUserData(user) not working?

**Review by remadyt — CHANGES_REQUESTED** [review]

File: src/components/shared/ProtectedRoute/ProtectedRoute.tsx [hunk 9]
```diff
@@ -3,17 +3,29 @@ import { FC, ReactElement } from 'react';
...
+
+    return <Navigate to={Paths.home} replace />;
+  }
+
+  if (isAuth) {
+    return <Navigate to={Paths.home} replace />;
   }
 
-  return isAuth ? <Navigate to={Paths.home} replace /> : (children as ReactElement);
+  return children as ReactElement;
```
Comment (remadyt) [inline src/components/shared/ProtectedRoute/ProtectedRoute.tsx:30]:
```
const ProtectedRoute = ({ children, isPrivate = true, allowedRoles }: ProtectedRouteProps) => {
  const { isAuth, user } = useAuth((state) => state);
  const userHasRequiredRole = user?.role !== undefined && allowedRoles.includes(user.role);

  if (!isAuth && isPrivate) {
    return <Navigate to={Paths.home} replace />;
 };
  if (isAuth && isPrivate && !userHasRequiredRole) {
    return <Navigate to={Paths.home} replace />;
};

  return children;
};
```
i think its same?

**Review by remadyt — CHANGES_REQUESTED** [review]

**Review by thelastandrew — APPROVED** [review]

**Review by remadyt — APPROVED** [review]

**Pull request merged**